src/libbson/src/bson/bson-vector.h

```diff
@@ -0,0 +1,591 @@
+{
+   size_t length = bson_vector_float32_view_length (view);
+   if (BSON_LIKELY (vector_offset_elements <= length && element_count <= length - vector_offset_elements)) {
+      size_t byte_offset = BSON_VECTOR_HEADER_LEN + vector_offset_elements * 4;
+#if BSON_BYTE_ORDER == BSON_LITTLE_ENDIAN
+      memcpy (view.binary.data + byte_offset, values, element_count * 4);
+#else
+      size_t i;
+      for (i = 0; i < element_count; i++) {
+         float aligned_tmp = BSON_FLOAT_TO_LE (values[i]);
+         memcpy (view.binary.data + byte_offset + i * 4, &aligned_tmp, 4);
+      }
+#endif
+      return true;
+   } else {
+      return false;
+   }
+}
+
+
+static BSON_INLINE bool
+bson_vector_packed_bits_const_view_read_packed (bson_vector_packed_bits_const_view_t view,
+                                                uint8_t *BSON_RESTRICT packed_values_out,
+                                                size_t byte_count,
+                                                size_t vector_offset_bytes)
+{
+   size_t length_bytes = bson_vector_packed_bits_const_view_length_bytes (view);
+   if (BSON_LIKELY (vector_offset_bytes <= length_bytes && byte_count <= length_bytes - vector_offset_bytes)) {
+      memcpy (packed_values_out, view.binary.data + BSON_VECTOR_HEADER_LEN + vector_offset_bytes, byte_count);
+      return true;
+   } else {
+      return false;
+   }
+}
+
+static BSON_INLINE bool
+bson_vector_packed_bits_view_read_packed (bson_vector_packed_bits_view_t view,
+                                          uint8_t *BSON_RESTRICT packed_values_out,
+                                          size_t byte_count,
+                                          size_t vector_offset_bytes)
+{
+   return bson_vector_packed_bits_const_view_read_packed (
+      bson_vector_packed_bits_view_as_const (view), packed_values_out, byte_count, vector_offset_bytes);
+}
+
+static BSON_INLINE bool
+bson_vector_packed_bits_view_write_packed (bson_vector_packed_bits_view_t view,
+                                           const uint8_t *BSON_RESTRICT packed_values,
+                                           size_t byte_count,
+                                           size_t vector_offset_bytes)
+{
+   size_t length_bytes = bson_vector_packed_bits_view_length_bytes (view);
+   if (BSON_LIKELY (vector_offset_bytes <= length_bytes && byte_count <= length_bytes - vector_offset_bytes)) {
+      if (byte_count == length_bytes - vector_offset_bytes && byte_count >= 1u) {
+         // This write touches the last byte in the vector:
+         // special-case that byte so we can ensure unused bits remain set to zero.
+         size_t other_bytes = byte_count - 1u;
+         memcpy (view.binary.data + BSON_VECTOR_HEADER_LEN + vector_offset_bytes, packed_values, other_bytes);
+         view.binary.data[BSON_VECTOR_HEADER_LEN + vector_offset_bytes + other_bytes] =
+            ((uint8_t) 0xFF << bson_vector_packed_bits_view_padding (view)) & packed_values[other_bytes];
```

Comment:
```suggestion
            (UINT8_C (0xFF) << bson_vector_packed_bits_view_padding (view)) & packed_values[other_bytes];
```
Minor suggestion to avoid the explicit cast.